### DISCUSSION TITLE
Using username in a TLS context is inappropriate

### DISCUSSION THREAD

**Discussion by dblas** [rabbitmq/rabbitmq-server#9226] [Other]
### Is your feature request related to a problem? Please describe.

Authentication is mainly based upon username (at least for MQTT). Ok, fine.
In TLS context you don't need username since you have the certificate to authenticate the device. 
Then, AFAIK, MQTT is made for connecting millions of objects.
How can you imagine managing millions of usernames?  
And what for? To separate queues?
That's insane!
That's not at all the way it should work, above all in a TLS context where certificate is used AS the main authenticator.
In this case you'll have 3 databases to manage: the usernames' one, the clientIDs' one and the certificates' one (serial number) which is managed by the CA.
10k objects => 30k items to manage! Yes, that's insane.
In MQTT, at least, the id used for managing client is the ... clientID. That the ID used for managing client sessions.
And, as you may know, you can perfectly separate queues through ACLs using %c in order for one client to not be able to see other client's queues.

### Describe the solution you'd like

In a **TLS context**, be able to extract the clientID from the certificate as it's possible to do it with username.
Thus, 
CN (for example) => clientID
and SAN (email, DNS, whatever) => username (which is already feasible as of current beta).

This way, it's then possible to separate queues using the /%c/ trick.
No need to manage thousands of usernames, clientIDs are dynamically managed by the Server and we only have to manage serial numbers through the CA. Only one database and every one is happy.
Authentication can still occur against the username but this one can be unique for a pool of hundreds of devices.

### Describe alternatives you've considered

Can't see other since managing thousands of usernames is insane if there is an alternative to separate queues.

**Comment by michaelklishin** [comment]
@dblas using highly emotional pitches will not work very well in this community. Instead of throwing around words like "insane" and asking questions that seem rhetorical to you but may or may be obvious to others, consider explaining what is it that you would like to see without judgement calls.

Or simply contribute a PR that would demonstrate it.

Throwing around words like "insane", "not at all how it should work" is great for letting steam out but not so much for improving open source software you very likely get entirely for free.

For x.509 certificate-based authentication scenarios, there is [a built-in plugin](https://github.com/rabbitmq/rabbitmq-server/tree/main/deps/rabbitmq_auth_mechanism_ssl) that makes it possible to extract a username from the most common certificate fields.

On top of that, you don't have to use the internal database for users. [LDAP](https://rabbitmq.com/ldap.html) or [any HTTPS-based service with a tiny API](https://github.com/rabbitmq/rabbitmq-server/tree/main/deps/rabbitmq_auth_backend_http) allow you to manage bajillions of usernames however you please.

**Comment by michaelklishin** [comment]
@dblas I will be very blunt here: if this discussion continues like this, it will not get very far, and we won't let insults towards the core team slide.

Please take a few steps back and describe what is it that you're trying to do, why, what authN/authZ backends (and mechanisms) have you considered.

How MQTT client information is extracted from x.509 certificates is not a recent development. I don't think 3.13 changes anything. So somehow the current set of options available works for at least some users.

There's still about 7-8 weeks before 3.13.0 ships, and months before 4.0 does. Reasonable changes can be introduced if instead of spitting emotions and making bold claims ("That's not at all the way it should work") someone presents are specific set of changes without calling the core team insane.

**Reply by dblas** [reply]
I wrote an answer this morning but it doesn't seem to have gone through.
Whatever.

Hello,

I think there is a little misunderstanding there. I do not blame anyone nor I insult anyone. I'm simply raising an operational problem that, I suppose, takes its root in a misunderstanding about the identity management part of a MQ broker whatever it is: Mosquitto, RabbitMQ, VerneMQ, HiveMQ, EMXQ.

None of the brokers I had time to look at (RabbitMQ, VerneMQ, Mosquitto) are able to extract the clientid from a certificate.
I think there is a reason but which one?

At this moment I think that's because too much confidence was placed into the id/mdp working that prevails in many applications today. I understand, but as soon as you're using certificates, things change: authentication is then made against a CA and doesn't need to be done against an LDAP or equivalent. That'd be a double-use.

Doing both is a lack of time as well as resource-hungry. You can but that will be complicated to keep things synchronized.
When we talk about human beings bearing a certificate that's understandable to carry on with username since you'll have many applications against which the user will have to authenticate and which do not support certificate.

But in a restricted context as such (I)IoT is, that'll be far TOO complicated. We're talking about machines, automates, not humans beings.

Deeply sorry if I sounded incorrect. That wasn't intended. I always write as I think.
For your information, I made the same request for Mosquitto and will do it for VerneMQ.

db

**Comment by michaelklishin** [comment]
Now, to comment on some specific bits that can be deciphered from the original post:

> CN (for example) => clientID

I am not convinced that a lot of MQTT tools do this by default. Client ID extraction instead of
using whatever the client provides could be a reasonable feature request but you make it sound
like the obvious solution that every other MQTT server has adopted. That's not the case. Client-provided IDs
are used very widely.

> SAN (email, DNS, whatever) => username

Username extraction from Subject Alternative Name has been supported for years.

> it's then possible to separate queues using the /%c/ trick

We'd need a clarification on what the well known "/c%/ trick" is. We do not guess in this community.

On the surface, it looks like the MQTT plugin is missing two things:

 * A way to extract client ID from a client certificate field, and ignore the client-provided value
 * Some "trick" that would make queue naming generation use a different value from what it does today?

Whether the plugin should use separate queues or not is an entirely separate discussion. Queues only offer destructive consumption, so either the plugin would need to support streams somehow (a change way out of scope for 3.13), or the use of separate queues with routing to several of them will be unavoidable in many cases. I don't see how this is relevant to how "insane" is what the plugin does w.r.t. username extraction.

**Reply by dblas** [reply]
> Now, to comment on some specific bits that can be deciphered from the original post:
>
>     CN (for example) => clientID
>
> I am not convinced that a lot of MQTT tools do this by default.

You are right. No MQTT brokers do this. Only some platforms (AWS, etc). 
My request was a feature request for RabbitMQ not a statement or the result of a survey.

And I don't understand why it's as it. In MQTT it seems to me far more logical to use the clientID as an identifier than to use the username. 
Maybe I've not the same mind logic - to avoid headaches - as most users have. Whatever ...

For me the ClientID is the true identifier and the session is based upon it. username can have its own use cases (to gather devices, to authenticate humans beings on the web interface, to create customer families) but clearly, if you have thousands of devices, you won't want to manage thousands of usernames whereas, if you are in a TLS context, you already manage thousands of certificates (serial numbers).
>
> Client ID extraction instead of
> using whatever the client provides could be a reasonable feature request but you make it sound
> like the obvious solution that every other MQTT server does. That's not the case. Client-provided IDs
> are used very widely.
>
>     SAN (email, DNS, whatever) => username
>
> Username extraction from Subject Alternative Name has been supported for years.
>
>     it's then possible to separate queues using the /%c/ trick
>
> We'd need a clarification on what the well known "/c%/ trick" is. We do not guess in this community.

%c and %u are the patterns used in VerneMQ for substituing respectively clientID (^{client_id}-.*) and username (^{username}-.*).

Sorry, I'm new to RabbitMQ, not accustomed to concepts yet. But, it's a product chosen by one of our subsidiaries and, as a group CISO, I have to specify a few rules to avoid wild usages.
>
> On the surface, it looks like the MQTT plugin is missing two things:
>
>     A way to extract client ID from a client certificate field, and ignore the client-provided value
>
YEP that's it. That simple said like this.
Are we sure there won't be any other consequences on other flows (authentication, etc)?
>
>     Some "trick" that would make queue naming generation use a different value from what it does today?
>
If I understand right, these tricks already exist: ^{client_id}-.* and ^{username}-.* I didn't see them before my post. Sorry.

The idea is to use the client_id pattern into ACL rules.

Thus, writing that devices can only have access to /sthing1/sthing2/^{client_id}-.*/tpc1/tpc2/ means each devices is confined to its own queue in only one sentence. No need to write as many ACL rules as they are devices.


> Whether the plugin should use separate queues or not is an entirely separate discussion. Queues only offer destructive consumption, so either the plugin would need to support streams somehow (a change way out of scope for 3.13), or the use of separate queues with routing to several of them will be unavoidable in many cases. I don't see how this is relevant to how "insane" is what the plugin does w.r.t. username extraction.

Username management insanity is related to operational constraint in a TLS context when you have thousands of devices: you'll have to manage them in LDAP, a database, whatever, whereas on another side you already manage devices serial numbers inside the AC that delivers the certificates.

Double-management, synch needed => headaches 
Of course you could create many usernames in advance during the deployment phase but sincerely that's not clean at all and leads to this insanity landscape. Sorry for that.

I'm convinced that's a very good way to make RabbitMQ go along with the new MQTT engine that came with version 3.12.
But, reading you, I'm afraid we'll have to explain why to many current users.
db

**Reply by michaelklishin** [reply]
The only reason I can think is, client IDs are not used for much more than logging in a lot of MQTT installations. Plus to enforce "connection uniqueness" as per the MQTT spec.

We can add extraction of client ID from a certificate field if present. How do you suggest that it is used then? To override the client connection-provided value? Something else?

**Reply by michaelklishin** [reply]
Another thing that I'd need to test is whether we pass client ID to authN backends. If we do, and the extraction from the certificate is done, a [tiny HTTP service you can develop](https://github.com/rabbitmq/rabbitmq-server/tree/main/deps/rabbitmq_auth_backend_http) would be able to authenticate clients based on their certificate-provided client IDs.

**Reply by dblas** [reply]
Having thinking about, yes, It seems relevant to push the clientID, once retrieved, into the same flows username follows (auth, authz). We don't know at this time what will be the usages. 
For me it makes sense (of course) to use the clientID above all as an authenticator. As you said, its uniqueness is, in principle, garanteed which makes it a far better candidate for authentication and authorization workflows than username.
db

**Comment by ansd** [comment]
The MQTT spec differentiates between client ID and user name. The client ID must always be unique. The client ID is used to track the state of an MQTT session. In contrast, the user name is meant to be used for authentication and authorization.

I'm not 100% sure as to why the distinction between client ID and user name was made in the MQTT spec, but the most plausible explanation to me is that clients can re-use the same username for authentication and authorization while still having a unique client ID to differentiate the different sessions.

When the client authenticates with a certificate, couldn't you use the same user name in the SAN of the different certificates? This is already supported today by the RabbitMQ MQTT plugin and solves your problem to not maintain N different user names. Obviously, the clients will still connect with different client IDs in the CONNECT packet.

> In TLS context you don't need username since you have the certificate to authenticate the device.

Authentication is not enough. You still need a username for RabbitMQ to perform authorization, right?

Extracting a client ID from the certificate doesn't make sense to me. The client ID is mandatory (albeit it could be empty) and already sent in the CONNECT packet from client to server.
I'd prefer to keep the client ID as a means to track session state, and the user name as a means to perform authentication and authorization.

**Reply by dblas** [reply]
Yep.
We can't see, I think, the problem under the same angle.
MQTT was specified in order to manage millions of devices. Right?
I can't see how I can manage millions of devices using a username by device. 
Sorry.
Even if an LDAP or a database or even a flat file can contain as many records I can't manage millions of records on one side (the Certificates Authority) and millions of records on the other side for authentication and authorization. That's a double-job.
And I don't want, of course, that a device be able to see in another's object queue.

Thus, either I decide that any username that comes from a certificate is, from the start, authenticated and authorized and I split apart queues using ^{username}-.* pattern but that's not the way MQTT sees such thing. Else I find another strategy.
And this strategy is to use the deviceid that can be part of the certificate (the CN for example), I split apart queues using the ^{client_id}-.* pattern and every device uses the same username. I've then only one username to manage. End of my job on the MQTT side.
I can also use different username (only a few ones) to distinguish between devices populations if needed.
In that case I could use the SAN (email type) for that if I do not want to intervene into the MQTT clients to set the username: only the certificate is to be considered and as it's changer every 2 years, the username can change at the same time.

Ok, no matter, I'll do the job myself :)

db

**Reply by ansd** [reply]
> I can't see how I can manage millions of devices using a username by device.
Sorry.

I ask again the same question: What prevents you to use the same user name in the SAN field for all your million devices?

**Reply by dblas** [reply]
Oups, sorry.
Because, as I wrote, I DO NOT want a device be ABLE to access another device's queue. 
One device = one client. That'd be a major flaw.
That's my first need.
I then need at least one UNIQUE "id" to split them apart.
RabbitMQ only proposes 2 "ids": username and clientid.
I'm not sure it's possible to get more whatever the context is.

I do not want to user the username for reasons I explained - too much job, double job -.

If I use the same username AND that's the only way I can identify the device THROUGH the certificate, that doesn't work.

One more thing: by definition the security is device-centric (the certificate is issued BY device and not by group of devices). Normal, we do not want another device to be comprised if one is compromised even if they belong to the same customer.

Hence the clientID as the major id. At the expense of the username.

db

**Reply by ansd** [reply]
RabbitMQ today supports variable expansion for topic permissions.
When an MQTT client subscribes with a topic filter, a new binding is created in RabbitMQ.
The user needs to have **read** access to the topic in order for the binding to be created (see [here](https://github.com/rabbitmq/rabbitmq-server/blob/0468dac975467fa4573c992a4cbcd661d33fd0ea/deps/rabbitmq_mqtt/src/rabbit_mqtt_processor.erl#L1042)).

What I wonder is:
Can you structure your topics in such a way that a device is only allowed to receive from a topic tree that contains its MQTT client ID?
You can then use the same username in the SAN field of all client certs (such that you have a single user object in RabbitMQ) while still ensuring that every device can only consume messages from the subset of the topic tree that contains its MQTT client ID. This kind of fulfils your requirement:
> Because, as I wrote, I DO NOT want a device be ABLE to access another device's queue. 

To provide an example:
Let's assume you have two (instead of millions of) devices. They are both issued separate certs, but the two certs contain the same user name in the SAN field.
For that single user name you create a single topic permission that is allowed to read from `^{client_id}.*`. If the client ID of device 1 is `device1`, it will then be only allowed to receive messages for topics that start with `device1` and, likewise, if the client ID of device 2 is `device2`, it will then only be allowed to receive messages for topics that start with `device2`.

This way you can have millions of devices, a single user name and a single topic permission while still ensuring that every device is allowed to only receive messages for its own topic subtree.

**Reply by dblas** [reply]
Yep, that's the way it works.
But where do you retrieve the client_id in order to use it into an ACL and to check the ACL against the client_id? I mean, in a TLS context.
Of course, you can always retrieve it through the MQTT protocol but you are then not sure it's not forged.
db

**Reply by ansd** [reply]
> but you are then not sure it's not forged

That's a good point, yes. Right now, all authenticated clients can choose whatever client ID they want.

If we were to add support for extracting the MQTT client ID from a client certificate, what would be the MQTT user name? I.e. which RabbitMQ user object would be used in order to perform the authorisation checks in https://www.rabbitmq.com/access-control.html#authorisation (checks such as "Is the connection allowed to read from / write to an exchange? Is the connection allowed to read from / write to a queue"?) Where is this user name extracted from if it doesn't come from the client certificate?

Or, are you suggesting that the client ID from the client cert is the same as the user name? If that's the case won't we again end up with millions of user objects in RabbitMQ?

**Reply by ansd** [reply]
Something which does make more sense to me is to extract both user name and client ID from the certificate.
E.g. client ID in the common name and user name in the SAN field or vice versa.
Then, we can still end up with a single user in RabbitMQ and enforce checks such as a specific MQTT client can only receive messages published to its own topic sub tree.

**Reply by ansd** [reply]
> For your information, I made the same request for Mosquitto and will do it for VerneMQ.

@dblas please provide links. I'm also interested in their responses.

**Reply by dblas** [reply]
In the case where a company develops devices for customers and will manage them, which is our case, this company will have the authority to design the clientID structure as it wants. Can be under the form of a mac address for example.

The MQTT username can be whatever the previous company wants, don't care. That's no more important if the clientID is used.

The username HAS to come from the certificate in order to follow the RabbitMQ authN flows. But, as it's already possible today, it can come from the SAN in order to avoid to program the device for that.

No. 1 username is enough, the same for all the devices. Or maybe 10 or 20 in order to classify them to be neat.
BUT as many clientIDs as there are devices.

The global process to generate certificates will take care of that: 
1st. clientID per device maybe based upon the first Ethernet interface's MAC Address, as such it'll impossible to have 2 identical clientIDs, this procedure is therefore beyond the business scope, and 
2nd. the username will be questioned in the RA form (RA = Registering Authority) with a default value and this RA form is filled up by the business (the department that orders the devices for the customers).
db

**Reply by dblas** [reply]
Mosquito: https://github.com/eclipse/mosquitto/issues/2886
VerneMQ: https://github.com/vernemq/vernemq/issues/2193
db

**Reply by ansd** [reply]
Right, so, in your case the user name could be the customer name. So, you have as many RabbitMQ user objects as customers.
And then each device has its MAC address as client ID.
Both client ID (MAC address), and user name (customer name) will be put somewhere into the client certificate and signed.
Then, we end up with only as many user objects in RabbitMQ as there are customers and you can still make use of `{client_id}` topic variable permissions to ensure that every device can only receive from its own topic subtree.
Such an approach makes sense to me.
However, only extracting the client ID, but not the user name from the certificate, makes less sense to me.
(Of course you could alternatively have separate RabbitMQ vhosts per customer.)

**Reply by dblas** [reply]
More or less. Don't think the username can be associated with the customer.
That's not a good example because customers won't never have access to the RabbitMQ interface, not even us: it's part of a contract because RabbitMQ is managed. And the lesser have access the better it is.
I was more thinking about world regions or countries or our subsidiary that managed the customers' contract (on a business point of view). In the latest case that would lead to only one username.

"However, only extracting the client ID, but not the user name from the certificate, makes less sense to me."
So do I.
Because I'd prefer to have the same process to set up everything that looks like an identifier and not having an id set through the certificate and another one set through the gitlab CI/CD that will be deployed on the device itself as part of its upgrade.
That's nonsense. I'm talking about security and, in this case, security is ensured through the CA not through the apt update.
db

**Reply by ansd** [reply]
I'm not sure I follow you with GitLab and apt update, but as long as we agree about the following, I'm happy:

RabbitMQ should be able to extract the following from the client certificate:
* MQTT user name only, or
* Both MQTT user name and MQTT client ID. If the client ID is extracted from the client certificate, the client ID sent in the CONNECT packet must be ignored by RabbitMQ.

The MQTT user name will continue to map to a user object in RabbitMQ.
Both user name and client ID will be propagated to the RabbitMQ authn / authz backends.

What should **not** work:
* Extract only the client ID from the certificate, but not the user name (because only the MQTT user name will map to a RabbitMQ user object and must therefore be provided in some way).

**Reply by ansd** [reply]
Does above fulfil your needs @dblas ?

**Reply by dblas** [reply]
Gitlab and apt are just the way the devices are developed and updated. Like a standard Linux machine.
db

**Reply by dblas** [reply]
Hello,
Discussing with someone it told me that whatever the context is, the ACLs don't exist outside a username environment.
I don't have time right now in order to corroborate or not this assertion but does that mean that we need at least 1 user in order to make an ACL like /.../{client_id}/... relevant and that if another username exist, the previous ACL will have to be copied in this username's environment?
db

**Reply by michaelklishin** [reply]
If you use the HTTP authN/authZ backend that was recommended several times above, you can ignore the username entirely in your service's implementation (and return a constant value that will be displayed everywhere). Not having a user on a connection is not an option in RabbitMQ, by design. Even MQTT's anonymous mode is backed by a user (the default user configurable for the plugin).

**Reply by michaelklishin** [reply]
Using client ID instead of username in queue names can also be an option if it does not affect the publishing or delivery path (that is, if we need to generate the name once when a client subscribes).

Introducing such optionality on the hot code paths will make the plugin less efficient for everyone, including those who do not have any of the needs expressed in this thread.

This increasingly sounds like a request to re-purpose client ID for username. I am not convinced that's how most MQTT implementations approach access control. The spec certainly does not suggest that client IDs can entirely replace usernames.

**Reply by dblas** [reply]
I don't think so since who talks about username will need to manage it (both locations as I explained).
Please consider the clientId as we discuss and let the users choose if they want to use the username (and manage it into the AC as well as into something else) => AuthN or if they want to use the clientID (no need to manage except into the AC) => AuthN.
db

**Reply by michaelklishin** [reply]
I'm not sure I understand what specifically you are asking for @dblas. "Use" for what specifically?

**Reply by dblas** [reply]
Use?
Did you read the whole thread, discussion with David in particular?
I was thinking my objectives were clear why I privilege the use of clientID rather than the username. 
I explain why it's inefficient and a waste of time to use username when you have thousands of devices which have, of course, already a unique id by themselves: materialized by the clientID. 
That's valid for MQTT and could be different in AMPQ (didn't look so far).
db

**Comment by michaelklishin** [comment]
https://github.com/rabbitmq/rabbitmq-server/issues/9323 and https://github.com/rabbitmq/rabbitmq-server/issues/9322 are two outcomes of this discussion.

#9322 is a no-brainer. #9323 does not make much sense given that

 * Client ID is not a means of authentication, although it potentially can be used as an extra factor in authN decision making
 * In 3.12, [client ID is already used in queue names](https://blog.rabbitmq.com/posts/2023/03/native-mqtt/#new-mqtt-qos-0-queue-type) together with QoS level
 * With an external HTTP-backed service, assuming that #9322 is done, the user will have a lot of flexibility when it comes to authN decisions

but we may add it anyway. There can be some value in encoding as much information as possible into the certificate.
